[DOCU-2823] New 3.1.x Upgrade doc (#4876)
Kong/docs.konghq.com: src/gateway/upgrade-31x.md

---
title: Upgrade Kong Gateway 3.1.x
---

Upgrade to major, minor, and patch {{site.base_gateway}}
releases using the `kong migrations` commands.

You can also use the commands to migrate all {{site.base_gateway}} open-source entities
to {{site.base_gateway}} (Enterprise). See
[Migrating from {{site.ce_product_name}} to {{site.base_gateway}}](/gateway/{{page.kong_version}}/migrate-ce-to-ke/).

If you experience any issues when running migrations, contact
[Kong Support](https://support.konghq.com/support/s/) for assistance.

## Upgrade path for {{site.base_gateway}} releases

Kong adheres to [semantic versioning](https://semver.org/), which makes a
distinction between major, minor, and patch versions.

The upgrade to {{page.kong_version}} is a **minor** upgrade.
The lowest version that Kong {{page.kong_version}} supports migrating from is 3.0.x.

{:.important}
> **Important**: Blue-green migration in traditional mode for versions below 2.8.2 to 3.0.x is not supported.

> The upcoming 2.8.2 release will include blue-green migration support. If you want
> to perform migrations for traditional mode with no downtime,
> upgrade to 2.8.2, [then migrate to {{page.kong_version}}](#migrate-db).

Reviewer: ```suggestion
The 2.8.2 release includes blue-green migration support. If you want
to perform migrations for traditional mode with no downtime,
upgrade to 2.8.2, [then migrate to {{page.kong_version}}](#migrate-db).
```

on the configuration templates, using `-w` for greater readability.

Here's how to see the differences between previous versions and 3.0.x:

```
git clone https://github.com/kong/kong
cd kong
git diff -w 2.0.0 3.0.0 kong/templates/nginx_kong*.lua
```

Adjust the starting version number (2.0.0 in the example) to the version number you are currently using.

To produce a patch file, use the following command:

```
git diff 2.0.0 3.0.0 kong/templates/nginx_kong*.lua > kong_config_changes.diff
```

Adjust the starting version number to the version number (2.0.0 in the example) you are currently using.

{% endnavtab %}
{% navtab Enterprise %}

The default template for {{site.base_gateway}} can be found using this command
on the system running your {{site.base_gateway}} instance:
`find / -type d -name "templates" | grep kong`.

When upgrading, make sure to run this command on both the old and new clusters,
diff the files to identify any changes, and apply them as needed.

{% endnavtab %}
{% endnavtabs %}

## Upgrade from 1.x/2.x to 3.1.x

### Traditional mode

Due to the changes in 3.1.x, you must upgrade from 2.8.2 or later to 3.0.x, then 3.1.x.

> {:.note}
> > These instructions will only work once 3.1.x is available.

Reviewer: ```suggestion

```

7. Start the 3.0.x cluster.
8. Now both the old (2.8.2) and new (3.0.x)
   clusters can now run simultaneously. Start provisioning 3.0.x nodes.
9. Gradually divert traffic away from your old nodes, and into
   your 3.1.x cluster. Monitor your traffic to make sure everything
   is going smoothly.
10. When your traffic is fully migrated to the 3.0.x cluster,
   decommission your old nodes.
11. Clone your database.
12. Download 3.1.x, and configure it to point to the cloned data store.
   Run `kong migrations up` and `kong migrations finish`.
13. Start the 3.1.x cluster.
14. Now both the old (3.0.x) and new (3.1.x)
   clusters can now run simultaneously. Start provisioning 3.1.x nodes.
15. Gradually divert traffic away from your old nodes, and into
   your 3.1.x cluster. Monitor your traffic to make sure everything
   is going smoothly.
16. When your traffic is fully migrated to the 3.1.x cluster,
   decommission your old nodes.

### Hybrid mode

Rolling upgrades from 1.x or 2.x to 3.1.x for hybrid mode are not supported.
To upgrade to 3.1.x, you must first [upgrade to 3.0.x](/gateway/3.0.x/upgrade).

## Upgrade from 3.0.x to 3.1.x

### Traditional mode

> {:.note}
> > These instructions will only work once 3.1.x is available.

Reviewer: ```suggestion

```
